### PR TITLE
[network/ebpf] https enable arm64 on kernel >= 5.5.0

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

https enable arm64 on kernel >= 5.5.0 with runtime compilation enabled

### Motivation

enable https monitoring on arm64

### Describe how to test/QA your changes

```
system_probe_config:
  log_level: debug
  enable_runtime_compiler: true
network_config:
  enable_http_monitoring: true
  enable_https_monitoring: true
```

 * run system-probe on arm64 with a kernel >= 5.5.0 ; like debian11 or ubuntu 21.10 or ami 5.10
 * `wget https://httpbin.org/anything/foo/200`
 * `sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections | tee cnx | jq '.conns[].tags , .tags'`
You should see openssl tags : `tls.library:openssl`
 * `sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/debug/http_monitoring`

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: test/kitchen/site-cookbooks/dd-system-probe-check/recipes/linux.rb [hunk 1]
```diff
@@ -50,6 +50,10 @@
 
 package 'socat'
 
+package 'wget'
+
+package 'curl'
```
Comment (brycekahle) [inline test/kitchen/site-cookbooks/dd-system-probe-check/recipes/linux.rb:55]: Are these related to this PR?

File: pkg/ebpf/c/bpf_helpers.h [hunk 2]
```diff
@@ -84,6 +84,17 @@ static int (*bpf_probe_read_user)(void* dst, int size, void* unsafe_ptr) = (void
 static int (*bpf_probe_read_kernel)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read_kernel;
 #endif
 
+/* bpf_probe_read_from_user is an helper to use bpf_probe_read_user() on kernel >= 5.5.0
+ * but will use bpf_probe_read() on older and prebuild build.
+ */
+#ifdef DD_PREBUILD
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/ebpf/c/bpf_helpers.h:90]: I think you could do something simpler, and more transparent like:
```
#if defined(DD_PREBUILD) || LINUX_VERSION_CODE < KERNEL_VERSION(5, 5, 0))
static int (*bpf_probe_read_user)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read;
static int (*bpf_probe_read_kernel)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read;
#else
static int (*bpf_probe_read_user)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read_user;
static int (*bpf_probe_read_kernel)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read_kernel;
#endif
```
and then use `bpf_probe_read_user` everywhere rather than a new function.

[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/ebpf/c/bpf_helpers.h:90]: Alternatively, we can try upstreaming a transparent replacement at the loader level in `cilium/ebpf` like what `libbpf` does.

File: releasenotes/notes/trace-agent-enable-https-monitoring-on-arm64-cc3dac41f9c0f28b.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    enable https monitoring on arm64 with kernel >= 5.5.0
```
Comment (maycmlee) [inline releasenotes/notes/trace-agent-enable-https-monitoring-on-arm64-cc3dac41f9c0f28b.yaml:11]:
```suggestion
    Enable https monitoring on arm64 with kernel >= 5.5.0,
```

File: releasenotes/notes/trace-agent-enable-https-monitoring-on-arm64-cc3dac41f9c0f28b.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    runtime compilation is require
```
Comment (maycmlee) [inline releasenotes/notes/trace-agent-enable-https-monitoring-on-arm64-cc3dac41f9c0f28b.yaml:12]:
```suggestion
    runtime compilation is require.
```

File: pkg/ebpf/c/bpf_helpers.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/ebpf/c/bpf_helpers.h:90]: cilium/ebpf load will fixup bpf_probe_read_{user,kernel} to bpf_probe_read if kernel doesn't support _{user,kernel}
so the fix will be to rename all our bpf_probe_read(), the loader will do the fixup https://github.com/cilium/ebpf/blob/master/linker.go#L196

[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/ebpf/c/bpf_helpers.h:90]: Oh nice! I didn't realize that was already there.

[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/ebpf/c/bpf_helpers.h:90]: Nevermind, they removed the check for `kern_version` in 5.0

File: releasenotes/notes/trace-agent-enable-https-monitoring-on-arm64-cc3dac41f9c0f28b.yaml (hunk 4 above)
Comment (brycekahle) [inline releasenotes/notes/trace-agent-enable-https-monitoring-on-arm64-cc3dac41f9c0f28b.yaml:12]:
```suggestion
    runtime compilation is required.
```

File: pkg/ebpf/c/bpf_helpers.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/ebpf/c/bpf_helpers.h]: I seen that libbpf do the same kind of fixup

File: pkg/network/tracer/tracer_test.go [hunk 5]
```diff
@@ -1628,12 +1646,16 @@ func TestHTTPStats(t *testing.T) {
...
 	if !httpSupported(t) {
 		t.Skip("HTTPS feature not available on pre 4.1.0 kernels")
 	}
-
-	if strings.HasPrefix(runtime.GOARCH, "arm") {
-		t.Skip("this feature is not yet support on arm")
+	if !httpsSupported(t) {
+		t.Skip("HTTPS feature not available on ARM pre 5.5.0 kernels")
+	}
+	if runningOnARM() && !cfg.EnableRuntimeCompiler {
```
Comment (brycekahle) [inline pkg/network/tracer/tracer_test.go:1657]: Can't we support prebuilt on ARM (>= 5.5.0) now that the loader does the helper translation for us?

Comment (nplanel) [reply pkg/network/tracer/tracer_test.go:1657]: yes, the issue is that offset-guessing is not working properly, I will take a look tomorrow

File: pkg/network/ebpf/c/prebuilt/http.c [hunk 6]
```diff
@@ -370,18 +385,15 @@ SEC("kprobe/do_sys_open")
 int kprobe__do_sys_open(struct pt_regs* ctx) {
     char *path_argument = (char *)PT_REGS_PARM2(ctx);
     lib_path_t path = {0};
-    bpf_probe_read(path.buf, sizeof(path.buf), path_argument);
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/ebpf/c/prebuilt/http.c:373]: bpf_probe_read_user() is basicaly copy_from_user_nofault() and return EFAULT if your outside of userspace boundary. If a EFAULT (page fault) happen the kernel will bzero(dst) :(

[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/network/ebpf/c/prebuilt/http.c:373]: Should we be using `bpf_probe_read_user_str` then?

[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/ebpf/c/prebuilt/http.c:373]: `bpf_probe_read_user_str` is supported only from 4.11 and there are no _str fixup/fallback

[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/network/ebpf/c/prebuilt/http.c:373]: Is it a matter of retrying in the face of `EFAULT`?

[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/ebpf/c/prebuilt/http.c:373]: I did a fallback as `bpf_probe_read_user` return us an error

File: pkg/network/ebpf/c/runtime/http.c [hunk 7]
```diff
@@ -372,20 +387,33 @@ int uprobe__gnutls_deinit(struct pt_regs* ctx) {
...
+    if (bpf_probe_read_user(path.buf, sizeof(path.buf), path_argument) >= 0) {
+// Find the null character and clean up the garbage following it
 #pragma unroll
-    for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
-        if (path.len) {
-            path.buf[i] = 0;
-        } else if (path.buf[i] == 0) {
-            path.len = i;
+        for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
+	    if (path.len) {
```
Comment (brycekahle) [inline pkg/network/ebpf/c/runtime/http.c:409]: weird indent here

File: pkg/network/ebpf/c/runtime/http.c [hunk 8]
```diff
@@ -372,20 +387,33 @@ int uprobe__gnutls_deinit(struct pt_regs* ctx) {
     return 0;
 }
 
-SEC("kprobe/do_sys_open")
-int kprobe__do_sys_open(struct pt_regs* ctx) {
+static __always_inline int fill_path_safe(lib_path_t *path, char *path_argument) {
+#pragma unroll
+    for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
+	bpf_probe_read_user(&path->buf[i], 1, &path_argument[i]);
```
Comment (brycekahle) [inline pkg/network/ebpf/c/runtime/http.c:393]: weird indent

File: pkg/network/ebpf/c/prebuilt/http.c [hunk 9]
```diff
@@ -366,20 +381,33 @@ int uprobe__gnutls_deinit(struct pt_regs* ctx) {
...
+    if (bpf_probe_read_user(path.buf, sizeof(path.buf), path_argument) >= 0) {
+// Find the null character and clean up the garbage following it
 #pragma unroll
-    for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
-        if (path.len) {
-            path.buf[i] = 0;
-        } else if (path.buf[i] == 0) {
-            path.len = i;
+        for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
+	    if (path.len) {
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/http.c:403]: weird indent

File: pkg/network/ebpf/c/prebuilt/http.c [hunk 10]
```diff
@@ -366,20 +381,33 @@ int uprobe__gnutls_deinit(struct pt_regs* ctx) {
     return 0;
 }
 
-SEC("kprobe/do_sys_open")
-int kprobe__do_sys_open(struct pt_regs* ctx) {
+static __always_inline int fill_path_safe(lib_path_t *path, char *path_argument) {
+#pragma unroll
+    for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
+	bpf_probe_read_user(&path->buf[i], 1, &path_argument[i]);
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/http.c:387]: weird indent

**Review by brycekahle — APPROVED** [review]
Some minor formatting issues. Otherwise 👍

**Review by lebauce — APPROVED** [review]

**Review by maycmlee — APPROVED** [review]
👍 for docs.
